### ISSUE TITLE
Unable to pause execution with root gdb

### ISSUE THREAD

**Issue by litist** [microsoft/vscode-cpptools#4243]
Type: Debugger

**Describe the bug**
- OS and Version: Ubuntu 16.04.6 LTS
- VS Code Version: 1.38.1
- C/C++ Extension Version: 0.25.1

We needed to debug a program which needs root privileges to be run. In order to do this a added a wrapper to gdb to lauch with sudo and without password.

```
#!/bin/bash
sudo gdb "$@"
```

When we launch a program with the wrapped dbg(set on miDebuggerPath), we are not able to pause the execution of the program or add and stop on a breakpoint.

Although, when the breakpoint is already set before lauching the debugger, it will pause there and we are able to step through the code and also set new breakpoints. After removing the breakpoints and continuing execution  we are again unable to pause execution.

**Comment by pieandcakes** [comment]
@litist Can you enable [engineLogging](https://code.visualstudio.com/docs/cpp/enable-logging-cpp#_enable-logging-for-the-debug-adapter) and see what gdb is returning as a message when we try and stop execution?

**Comment by litist** [comment]
When i try to add a breakpoint i get:
> 1: (16115) Send Event AD7BreakpointErrorEvent

while pressing pause gives on debug info and has no effect.
Pressing stop also gives no debug information, the debugger control is gone but the debugged program keeps running in the terminal.

**Comment by litist** [comment]
With traceRespone enabled i, get following on breakpoint setting:
> <--   C (setBreakpoints-9): {"command":"setBreakpoints","arguments":{"source":{"name":"tutorial.cxx","path":"/home/FE/lindstedt/mn/tutorial.cxx"},"lines":[15],"breakpoints":[{"line":15}],"sourceModified":false},"type":"request","seq":9}
--> E (output): {"type":"event","event":"output","body":{"category":"console","output":"1: (16711) Send Event AD7BreakpointErrorEvent\n"},"seq":345}
1: (16711) Send Event AD7BreakpointErrorEvent
--> R (setBreakpoints-9): {"type":"response","request_seq":9,"success":true,"command":"setBreakpoints","body":{"breakpoints":[{"id":1,"verified":true,"line":15}]},"seq":347}
--> E (breakpoint): {"type":"event","event":"breakpoint","body":{"reason":"changed","breakpoint":{"id":1,"verified":false,"message":"Attempting to bind the breakpoint....","line":15}},"seq":349}


Pause:
> <--   C (pause-10): {"command":"pause","arguments":{"threadId":26641},"type":"request","seq":10}
--> R (pause-10): {"type":"response","request_seq":10,"success":true,"command":"pause","body":{},"seq":352}

Stop:
> <--   C (disconnect-11): {"command":"disconnect","arguments":"restart":false},"type":"request","seq":11}

**Comment by pieandcakes** [comment]
@litist can you run it again with `engineLogging` set to `true`? The `traceResponse` log only shows what VS Code sends our debug adapter and what we respond with. I'm interested in what `gdb` is doing.

**Comment by litist** [comment]
> When i try to add a breakpoint i get:
> 
> > 1: (16115) Send Event AD7BreakpointErrorEvent
> 
> while pressing pause gives on debug info and has no effect.
> Pressing stop also gives no debug information, the debugger control is gone but the debugged program keeps running in the terminal.

This was the only output i got with engineLogging enabled.

**Comment by pieandcakes** [comment]
You should have:
```
"logging": { "engineLogging": true }
```

in your `launch.json` which will give you everything we send and receive to `gdb`. If you aren't seeing anything in the `debugConsole`, please post your configured `launch.json`.

**Comment by litist** [comment]
This is my configuration:
```
    "configurations": [
        {
            "name": "(gdb) sgdb",
            "type": "cppdbg",
            "request": "launch",
            "program": "${workspaceFolder}/a.out",
            "args": [],
            "stopAtEntry": false,
            "cwd": "${workspaceFolder}",
            "environment": [],
            "externalConsole": false,
            "MIMode": "gdb",
            "miDebuggerPath": "${workspaceFolder}/sgdb",
            "logging": { "engineLogging": true },
            "setupCommands": [
                {
                    "description": "Enable pretty-printing for gdb",
                    "text": "-enable-pretty-printing",
                    "ignoreFailures": true
                }
            ]
        }
```
and this is what i get in the debug console, after setting three different breakpoints.

> ...
1: (1761) ->(gdb)
1: (1762) 1019: elapsed time 1
1: (1763) Send Event AD7EntryPointEvent
1: (1765) <--exec-continue
1: (1765) ->^running
1: (1765) ->*running,thread-id="all"
1: (1765) ->(gdb)
1: (17518) Send Event AD7BreakpointErrorEvent
1: (19023) Send Event AD7BreakpointErrorEvent
1: (20687) Send Event AD7BreakpointErrorEvent

**Comment by pieandcakes** [comment]
@litist I think you cut out too much in the log. I'm interested in the breakpoint binding messages. If you don't want to post the log publicly, feel free to send me an email referencing this issue and I can look at it privately.

**Comment by joelspadin-garmin** [comment]
I'm seeing an issue with the same symptoms and log output, but in a completely different scenario.

* OS and Version: Ubuntu 18.04.3 LTS and Windows 7 (6.1.7601)
* VS Code Version: 1.40.0
* C/C++ Extension Version: 0.26.1

I'm working on a project which uses a gdbinit script to connect to a remote ARM Linux device via SSH. When I debug from a Windows machine, our toolchain has GDB version 8.0, and everything works correctly. However, when I debug from a Linux machine, our toolchain has GDB version 7.11.1 and I see the following behaviors:

- Breakpoints that are already set before starting debugging work.
- Attempting to pause a running program does nothing.
- Setting a breakpoint while the program is running does not work. I get a gray circle in the margin instead of a red one.
- When I stop debugging, the program remains running on the remote device instead of stopping.

If I run GDB directly in a terminal (not using MI mode), all of these things work as expected, so this seems to be some sort of issue with VS Code controlling either GDB 7.x or GDB on Linux in general.

This is what my launch configuration looks like:

```json
"configurations": [
        {
            "name": "gdb",
            "type": "cppdbg",
            "request": "launch",
            "program": "${workspaceFolder}/path/to/program",
            "cwd": "${workspaceFolder}",
            "targetArchitecture": "arm",
            "MIMode": "gdb",
            "miDebuggerPath": "/path/to/gdb",
            "miDebuggerArgs": "--command /path/to/gdbinit.py",
            "externalConsole": false,
            "customLaunchSetupCommands": [
                {
                    "text": "-target-select garmin --pause",
                    "description": "Use target defined in gdbinit.py to connect to the remote device"
                }
            ],
            "logging": { "engineLogging": true, "trace": true, "traceResponse": true }
        }
    ]
```

And here is what gets logged after each action I take. I've removed proprietary info from the logs. I may be able to provide more information privately if needed.

**Debugging from Linux machine (broken)**

Set breakpoint while program is running:

```
<--   C (setBreakpoints-22): {"command":"setBreakpoints","arguments":{"source":{"name":"source.c","path":"/path/to/source.c","sources":[],"checksums":[]},"lines":[1140],"breakpoints":[{"line":1140}],"sourceModified":false},"type":"request","seq":22}
--> E (output): {"type":"event","event":"output","body":{"category":"console","output":"1: (32688) Send Event AD7BreakpointErrorEvent\n"},"seq":1920}
1: (32688) Send Event AD7BreakpointErrorEvent
--> R (setBreakpoints-22): {"type":"response","request_seq":22,"success":true,"command":"setBreakpoints","body":{"breakpoints":[{"id":2,"verified":true,"line":1140}]},"seq":1922}
--> E (breakpoint): {"type":"event","event":"breakpoint","body":{"reason":"changed","breakpoint":{"id":2,"verified":false,"message":"Attempting to bind the breakpoint....","line":1140}},"seq":1924}
```

Remove the breakpoint that didn't correctly set:

```
<--   C (setBreakpoints-24): {"command":"setBreakpoints","arguments":{"source":{"name":"source.c","path":"/path/to/source.c","sources":[],"checksums":[]},"lines":[],"breakpoints":[],"sourceModified":false},"type":"request","seq":24}
--> R (setBreakpoints-24): {"type":"response","request_seq":24,"success":true,"command":"setBreakpoints","body":{"breakpoints":[]},"seq":1930}
```

Pause the program:

```
<--   C (pause-23): {"command":"pause","arguments":{"threadId":2},"type":"request","seq":23}
--> R (pause-23): {"type":"response","request_seq":23,"success":true,"command":"pause","body":{},"seq":1927}
```

Stop debugging:

```
<--   C (disconnect-36): {"command":"disconnect","arguments":{"restart":false},"type":"request","seq":36}
```

**Debugging from Windows machine (working correctly)**

Set breakpoint while program is running: [Windows set breakpoint.log](https://github.com/microsoft/vscode-cpptools/files/3838521/Windows.set.breakpoint.log)

Remove the breakpoint:

```
<--   C (setBreakpoints-23): {"command":"setBreakpoints","arguments":{"source":{"name":"source.c","path":"d:\\path\\to\\source.c","sources":[],"checksums":[]},"lines":[],"breakpoints":[],"sourceModified":false},"type":"request","seq":23}
--> R (setBreakpoints-23): {"type":"response","request_seq":23,"success":true,"command":"setBreakpoints","body":{"breakpoints":[]},"seq":2405}
```

Pause the program: [Windows pause program.log](https://github.com/microsoft/vscode-cpptools/files/3838523/Windows.pause.program.log)

Stop debugging: [Windows stop debugging.log](https://github.com/microsoft/vscode-cpptools/files/3838522/Windows.stop.debugging.log)

**Comment by pieandcakes** [comment]
Hi @joelspadin-garmin, 

Can you make sure that you have `engineLogging` set for the Linux scenario? I need the `gdb` logging. 

Can you also run: `-list-target-features` on the Linux `gdb` and let me know if `async` is available? 

Are you using `ssh` to run `gdb` on the ARM device? The `launch.json` is not showing enough information for me to duplicate this scenario. 

If you don't want to post here, my email address is in my profile.

**Comment by joelspadin-garmin** [comment]
`engineLogging` is already true in my `launch.json`. Did you want me to provide logs where that is the _only_ logging option set, or is there some logging you're expecting that isn't showing up?

> Can you also run: -list-target-features on the Linux gdb and let me know if async is available?

Both Linux and Windows versions of GDB output this, so it looks like async is not available.

```
(gdb) 
-list-target-features
^done,features=[]
```

> Are you using ssh to run gdb on the ARM device? The launch.json is not showing enough information for me to duplicate this scenario.

Yes. The custom `-target-select` command from my launch.json is implemented by the gdbinit script and does the following:

1. Opens an SSH connection to the remote device.
2. Pushes the latest program binary to the remote device.
3. Starts `gdbserver --once --multi localhost:<port>` on the remote device.
4. Runs `target extended-remote <address>:<port>` to connect GDB to the remote device.
5. Runs other commands to load symbols, etc.
6. Runs `set remote exec-file <path/to/program>`.
7. Runs `set scheduler-locking step` so stepping behaves more nicely.
8. Runs `set arm abi AAPCS` so GDB uses the proper ABI for the platform.

I also found #807, which looks like it may be the same issue or related.

**Comment by joelspadin-garmin** [comment]
I just now discovered that you probably meant I should run `-list-target-features` while the debugger is attached to the remote device. If I run `-exec -list-target-features` in VS Code's debug console, it looks like async **is** available.

```
-exec -list-target-features
result-class: done
features: [async]
```

I also tried compiling GDB 8.3.1 from source for ARM, and it behaved the same way as the version 7.11.1 that is part of our toolchain, so I don't think this is caused by using an old version of GDB.

**Comment by pieandcakes** [comment]
@joelspadin-garmin I was looking for the `gdb` output for the Linux scenario and all that is in the block above is the VSCode Protocol messages. I see now that the linked issue is the same problem. Let me see if i can track down a raspberry pi and try gdb <-> gdbserver.

**Comment by joelspadin-garmin** [comment]
Maybe useful new info: I previously stated that I was able to pause a program when running the debug session from Windows. That was not correct. Nothing happens when I press the pause button on either platform.

Here is a comparison of the behaviors I'm seeing on each platform:

Function | Windows Behavior | Linux Behavior
--|--|--
Pause | Nothing happens. | Nothing happens.
Set breakpoint before starting debugging | Works. | Works.
Set breakpoint while program is running | Works. | "Attempting to bind the breakpoint..." forever.
Stop debugging | Program terminates. | If program was paused in debugger, it terminates; otherwise it keeps running in the background.

Would it be helpful if I checked for differences in the logs between Linux and Windows?

**Comment by pieandcakes** [comment]
@joelspadin-garmin The logging that would give us the most info would be the `engineLoggging` setting as that tells us what we send to `gdb` and the responses/events we get back. 

To stop the debugger, we have two ways of doing it:
* `kill -2 <debuggeepid>`
* `-exec-interrupt` through `gdb`

If you have a separate `ssh` connection in and send `kill -2 <debuggee pid>`, will it show stopped in VS Code?

**Comment by joelspadin-garmin** [comment]
On Linux, with `engineLogging` enabled, there is no output in either the debug console or the internal cppdbg terminal when I attempt to pause or stop the debugger.

On Windows, there is also no output when I attempt to pause the program. However, stopping the debugger logs an `-exec-interrupt` command followed by lots of thread info messages, and eventually:

```
1: (24734) ->(gdb)
1: (24734) ->&"kill\n"
1: (24734) ->~"Kill the program being debugged? (y or n) [answered Y; input not from terminal]\n"
(many thread exited/destroy events)
1: (24763) ->(gdb)
1: (24764) 1053: elapsed time 909
1: (24770) <-logout
1: (24772) Send Event AD7ProgramDestroyEvent
The program 'C:\path\to\program' has exited with code 0 (0x00000000).
```


Running `kill -2 <pid>` in a separate ssh connection pauses the debugger in the main thread's event loop, and GDB logs something like...

```
Thread 1 "main" received signal SIGINT, Interrupt.
```

...so I think that is working as expected. This works on both Windows and Linux.

I also tried running GDB in MI mode manually (not from vscode) and `-exec-interrupt` appears to be working:

```
(gdb)
-gdb-set target-async on
^done
(gdb)
-target-select garmin --pause
=cmd-param-changed,param="remote exec-file",value="/path/to/program"
^connected
(gdb)
-exec-run
=thread-group-started,id="i1",pid="3412"
=thread-created,id="1",group-id="i1"
=library-loaded,id="/lib/ld-linux-armhf.so.3",target-name="/lib/ld-linux-armhf.so.3",host-name="/path/to/toolchain/lib/ld-linux-armhf.so.3",symbols-loaded="0",thread-group="i1",ranges=[{from="redacted",to="redacted"}]
^running
*running,thread-id="all"
(gdb)
many more =library-loaded messages...
-exec-interrupt
^done
(gdb)
~"\nProgram"
~" received signal SIGINT, Interrupt.\n"
=cmd-param-changed,param="scheduler-locking",value="step"
=cmd-param-changed,param="arm abi",value="AAPCS"
*stopped,reason="signal-received",signal-name="SIGINT",signal-meaning="Interrupt",frame=redacted
```

The only other difference between my Linux and Windows configurations is that I use `"externalConsole": true` on Windows and `false` on Linux. I don't really know what this affects, but I found that if I use an internal console on Windows it doesn't properly escape some paths and fails on an invalid escape sequence, and if I use an external console on Linux, it hangs and never starts debugging.

**Comment by joelspadin** [comment]
Someone mentioned that setting [this setting](https://visualgdb.com/KB/?ProblemID=breakinhang) fixed issues when using VisualGDB to debug the same hardware. Could vscode be trying to resume and interrupt the program quickly when connecting, and running into the same race condition?

**Comment by joelspadin-garmin** [comment]
#833 also looks like the same issue.

**Comment by joelspadin-garmin** [comment]
My issue may be caused by https://github.com/microsoft/MIEngine/blob/80e51dca0712b02910889d167ec356c4d0761279/src/MICore/Debugger.cs#L725. I have a remote target, but that's hidden away inside my custom target command so VS Code won't see it that way. If for some reason it's finding some process as the debuggee (which would be incorrect since the debuggee is running on another machine), it'll try to interrupt that instead of using `-exec-interrupt`.

Alternatively, it may be falling into the `-exec-interrupt` case but it isn't working for some reason. That would be strange because `-exec-interrupt` works when I run GDB manually, but VS Code responds with "Unable to perform this action because the process is running." when I try to run `-exec-interrupt` from the debug console, so it's possible that it's failing even though async is enabled.

Is there any documentation on debugging MIEngine with VS Code so I can figure out exactly what it's doing?

I'll also try to rework our gdbinit script so that I can have VS Code create a remote connection normally to see if that fixes things.

**Comment by pieandcakes** [comment]
```
That would be strange because -exec-interrupt works when I run GDB manually, but VS Code responds with "Unable to perform this action because the process is running." when I try to run -exec-interrupt from the debug console, so it's possible that it's failing even though async is enabled.
```

We give that error message when the debuggee is running and the command isn't sent to `gdb`.

```
Is there any documentation on debugging MIEngine with VS Code so I can figure out exactly what it's doing?
```

We have https://github.com/microsoft/vscode-cpptools/blob/master/Documentation/Debugger/How%20To%20Debug%20MIEngine.md . Let me know if you need more information.

If you attempt to use `miDebugServerAddress` to connect to the remote `gdbserver` does it do the right thing?

**Comment by joelspadin-garmin** [comment]
> If you attempt to use miDebugServerAddress to connect to the remote gdbserver does it do the right thing?

That's the next thing I plan to try. I'll need to figure out how to replicate what my gdbinit script does with vscode's settings first.

This probably should go on MIEngine's issues page, but it would be nice to have better support for user-defined targets created by gdbinit scripts.

**Comment by pieandcakes** [comment]
If you have some suggestions and leave them in that repo, I can take a look when I have time.

**Comment by joelspadin-garmin** [comment]
> If you have some suggestions and leave them in that repo, I can take a look when I have time.

I was looking through the MIEngine code and saw it only used `-target-select remote` and `-target-select core`, but then I remembered that `customLaunchSetupCommands` is a thing and I'm already using it. 🤦

I'll have to do more testing, but I *think* that simply setting `miDebugServerAddress` has fixed most of my issues without me having to manually set up vscode to launch the target. Thanks!

I don't know how much of this is relevant to @litist 's original issue though, since I'm using a custom launch command which tricked MIEngine into treating a remote connection like a local one, but they don't appear to be doing that. I'd guess that MIEngine is doing the wrong thing to interrupt their program, but for a different reason. Maybe MIEngine is sending a SIGINT but their program is somehow ignoring it?

**Comment by joelspadin-garmin** [comment]
...almost all of my problems. I can now set breakpoints while the program is running, but pressing the pause button does nothing until I set or stop at a breakpoint. Then it starts working again for the rest of that debugging session.

What function in MIEngine happens when you press the pause button? I can try debugging that sometime later.

**Comment by pieandcakes** [comment]
@joelspadin-garmin If you set a breakpoint [here](https://github.com/microsoft/MIEngine/blob/2bb9706607edf0ce839bab4eb131adc8a555b6ec/src/MICore/Debugger.cs#L573), you should be able to step.

**Comment by pmattione** [comment]
Was this ever resolved?  I have the same problem as the original poster: The debugger can't pause while running gdb with sudo via an indirection (gcc 10.2). This is directly in Ubuntu, nothing remote.  

Debug Console upon Pause:
 ```
<--   C (pause-9): {"command":"pause","arguments":{"threadId":50373},"type":"request","seq":9}
--> R (pause-9): {"type":"response","request_seq":9,"success":true,"command":"pause","body":{},"seq":829}
```

My /usr/bin/gdbsudo:
```
#!/bin/sh
sudo gdb $@
```

My launch.json:
```
"configurations": [
	{
		"name": "(gdb) MyProgram",
		"type": "cppdbg",
		"request": "launch",
		"program": "${workspaceFolder}/build/MyProgram",
		"args": [],
		"stopAtEntry": false,
		"cwd": "${workspaceFolder}",
		"environment": [],
		"externalConsole": false,
		"MIMode": "gdb",
		"miDebuggerPath": "/usr/bin/gdbsudo",
		"setupCommands": [
			{
				"description": "Enable pretty-printing for gdb",
				"text": "-enable-pretty-printing",
				"ignoreFailures": true
			}
		],
		"logging": {
			"engineLogging": true,
			"trace": true,
			"traceResponse": true
		 }
	}
```

VSCode version:
```
Version: 1.51.1
Commit: e5a624b788d92b8d34d1392e4c4d9789406efe8f
Date: 2020-11-10T23:31:29.624Z
Electron: 9.3.3
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Linux x64 5.4.0-54-generic snap
```

**Comment by pmattione** [comment]
You can even repro this with:

```
#include <chrono>
#include <thread>
#include <iostream>

int main(void)
{
	static constexpr std::chrono::seconds sSleepTime{ 1 };
	size_t loop_count = 0;
	while(true)
	{
		std::cout << loop_count++ << "\n";
		std::this_thread::sleep_for(sSleepTime);
	}
	return 0;
}
```

**Comment by pmattione** [comment]
Any help would be appreciated.  This is making it difficult to debug my web server, which needs to be run with admin privileges to use port 80.

**Comment by justanotheranonymoususer** [comment]
I'm having a similar problem on Windows with gdb:
I try to debug a GUI application, and it just doesn't work - I can step at the beginning, but if I run the program, I can no longer pause it, and placing breakpoints don't work. This makes the debugger non-functional. Please, any help is appreciated.

I also commented here in a similar ticket:
https://github.com/microsoft/vscode-cpptools/issues/5294

Please, if you have any workaround, it will help.

**Comment by OwlVision** [comment]
@justanotheranonymoususer , if question still relevant, i found some workaround, that helped me in similiar case. I am working with VSCode + MinGW : pause button and breakpoints ( under execution ) doesnt work too.

But if run application externally under gdbserver, and connect VSCode debugger to "remote" debug process - then pause button and breakpoints becomes functional. It works for me.

I saw somewhere scarce explanation of problem, like "its a Cygwin/MinGW debugger specificities, that incompatable with vscode-debug-wrapper", but I cant give a link, cause i've lost it.
I think, it relates to the way, that how gdb works and suspend processes ( sends interrupt on Ctrl+C shortcut in console ), and that vscode-debug-wrapper cant do this.

launch.bat

    %MINGW%/gdbserver.exe localhost:2000 app.exe

tasks.json _( fragment )_

    {
        "type" : "shell",
        "label": "Run application with gdbserver",
        "command" : "Start-Process -FilePath .\\launch.bat"
    }

launch.json

    {
        "name": "Remote Debug",
        "type": "cppdbg",
        "request": "launch",
        "program": "${workspaceFolder}\\app.exe",
        "stopAtEntry": false,
        "cwd": "${workspaceFolder}",
        "environment": [],
        "externalConsole": true,
        "MIMode": "gdb",
        "miDebuggerPath": "${env:MINGW}\\gdb.exe",
        "miDebuggerServerAddress": "localhost:2000",
        "setupCommands": [
             {
                 "text": "-enable-pretty-printing",
                 "ignoreFailures": true
             }
         ],
         "preLaunchTask": "Run application with gdbserver"
    }

**Comment by doug65536** [comment]
Closed automatically? What a load of crap.

**Comment by sean-mcmanus** [comment]
@doug65536 It was closed because of the tag for needing more info. Can you add more info or create a new issue with more info?

**Comment by pmattione** [comment]
I added plenty of info on Nov 26, 2020 and it was ignored.  This problem is still occurring today (in VSCode 1.71.2)

**Comment by sean-mcmanus** [comment]
@WardenGnaw I

**Comment by JasonDictos** [comment]
Running into this same problem since forever, anyone need any info?

```
--> R (pause-25): {"type":"response","request_seq":25,"success":true,"command":"pause","body":{},"seq":901}
<--   C (pause-26): {"command":"pause","arguments":{"threadId":72157},"type":"request","seq":26}
--> R (pause-26): {"type":"response","request_seq":26,"success":true,"command":"pause","body":{},"seq":904}
```

**Comment by Dimonsky** [comment]
I have the same problem when placing the breakpoint in VSCode, when gdg is running under sudo.
1) if "stopAtEntry" = true, the gdb breaks automatically in main(). At this moment you can place any breakpoint in the code. But as soon as you press F5 and let the program run, all later placed breakpoints are getting gray and never hit.

<--   C (setBreakpoints-42): {"command":"setBreakpoints","arguments":{"source":{"name":"ListOfTagsLocal.cpp","path":"/home....../file.cpp"},"lines":[39],"breakpoints":[{"line":39}],"sourceModified":false},"type":"request","seq":42}
--> E (output): {"type":"event","event":"output","body":{"category":"console","output":"1: (130617) Send Event AD7BreakpointErrorEvent\n"},"seq":1242}
1: (130617) Send Event AD7BreakpointErrorEvent
--> R (setBreakpoints-42): {"type":"response","request_seq":42,"success":true,"command":"setBreakpoints","body":{"breakpoints":[{"id":9,"verified":true,"line":39,"BoundBreakpoints":[]}]},"seq":1244}
--> E (breakpoint): {"type":"event","event":"breakpoint","body":{"reason":"changed","breakpoint":{"id":9,"verified":false,"message":"Attempting to bind the breakpoint....","line":39,"BoundBreakpoints":[]}},"seq":1246}

**Comment by consultsoped** [comment]
I have the exact same issue as @Dimonsky running vscode on WSL

```
> <--   C (pause-18): {"command":"pause","arguments":{"threadId":3734},"type":"request","seq":18}
> --> R (pause-18): {"type":"response","request_seq":18,"success":true,"command":"pause","body":{},"seq":394}
> <--   C (pause-19): {"command":"pause","arguments":{"threadId":3734},"type":"request","seq":19}
> --> R (pause-19): {"type":"response","request_seq":19,"success":true,"command":"pause","body":{},"seq":397}
> <--   C (setBreakpoints-20): {"command":"setBreakpoints","arguments":{"source":{"name":"main.c","path":"/mnt/c/work_git/HelloWorld/src/main.c","sources":[],"checksums":[]},"lines":[7,9],"breakpoints":[{"line":7},{"line":9}],"sourceModified":false},"type":"request","seq":20}
> --> E (output): {"type":"event","event":"output","body":{"category":"console","output":"1: (34632) Send Event AD7BreakpointErrorEvent\n"},"seq":400}
> 1: (34632) Send Event AD7BreakpointErrorEvent
> --> R (setBreakpoints-20): {"type":"response","request_seq":20,"success":true,"command":"setBreakpoints","body":{"breakpoints":[{"id":2,"verified":true,"line":7,"BoundBreakpoints":[]},{"id":1,"verified":false,"message":"Attempting to bind the breakpoint....","line":9,"BoundBreakpoints":[]}]},"seq":402}
> --> E (breakpoint): {"type":"event","event":"breakpoint","body":{"reason":"changed","breakpoint":{"id":2,"verified":false,"message":"Attempting to bind the breakpoint....","line":7,"BoundBreakpoints":[]}},"seq":404}
```

**Comment by Yemaoxin** [comment]
So? Where is the solution?

**Comment by MMRep** [comment]
Same issue as @litist.

I have to debug an application that requires root privileges.
I've modified the gdb path in launch.json (as mentioned in #861) to point to a script with following content:
```
pkexec /usr/bin/gdb "$@"
```

Stopping the debug session works, but pausing it doesn't work.
When I press "pause" following message is written to the debug console:
```
<--   C (pause-44): {"command":"pause","arguments":{"threadId":76318},"type":"request","seq":44}
--> R (pause-44): {"type":"response","request_seq":44,"success":true,"command":"pause","body":{},"seq":736}
```
I've also checked the thread ID `76318` and it is the process I am debugging.

The issue can be reproduced by simply making `miDebuggerPath` point to a gdb script that calls gdb via pkexec as shown above. A simple empty loop in main() is enough.

Maybe a more convenient solution for debugging applications that require root privileges would avoid this problem here?

**Comment by bjrxyz** [comment]
This issue has been around for five years and still seems unresolved. I'm left with no choice but to launch VSCode with root privileges to debug programs that require root access.

**Comment by jtousley** [comment]
@sean-mcmanus What is the status of this issue? Surprised to see an issue that is so clearly delineated and reproducible, affecting so many users, completely ignored for so many years. Can confirm this is still equally relevant in 2025.

Version: 1.101.2
Commit: 2901c5ac6db8a986a5666c3af51ff804d05af0d4
Date: 2025-06-24T20:27:15.391Z
Electron: 35.5.1
ElectronBuildId: 11727614
Chromium: 134.0.6998.205
Node.js: 22.15.1
V8: 13.4.114.21-electron.0
OS: Linux x64 6.11.0-29-generic

**Comment by JasonDictos** [comment]
> [@sean-mcmanus](https://github.com/sean-mcmanus) What is the status of this issue? Surprised to see an issue that is so clearly delineated and reproducible, affecting so many users, completely ignored for so many years. Can confirm this is still equally relevant in 2025.
> 
> Version: 1.101.2 Commit: 2901c5ac6db8a986a5666c3af51ff804d05af0d4 Date: 2025-06-24T20:27:15.391Z Electron: 35.5.1 ElectronBuildId: 11727614 Chromium: 134.0.6998.205 Node.js: 22.15.1 V8: 13.4.114.21-electron.0 OS: Linux x64 6.11.0-29-generic

Still, totally difficult to work around too! Some of us are writing back end daemons who never run as standard users, and noine of the debugging options that vscode provides makes this an easy task.

**Comment by sean-mcmanus** [comment]
@jtousley I work on the language service component. We don't have anyone working on the VS Code C/C++ debugger currently. The bug might be fixable via the open source MIEngine (which is being worked on, but I think their focus is on VS 2022+).
